### ISSUE TITLE
Camera2D does not work properly when added as child of a moving node with smoothing off and smoothing speed greater than 0

### ISSUE THREAD

**Issue by adeluiz** [godotengine/godot#16918]
**Godot version:**
3.0 stable

**OS/device including version:**
tested on win 10, android and osx

**Issue description:**
When you create a simple camera as a child of the player, just to follow it, if you have disabled the smoothing but there is a speed greater than 0, the sprite does some strange flickering

Seems related to https://github.com/godotengine/godot/issues/14545 but it has been closed without solution.

**Steps to reproduce:**
To isolate the bug I've create a simple project, with just a camara that follows the player. The player has an increased velocity with time and it goes increasingly off-centered. Just run it and watch. And after that, set the smoothing velocity to 0 and run it again (it goes centered as it should be).

**Minimal reproduction project:**
Due to the incorrect use of _proccess, I've uploaded a new one that moves a sprite

[Camera2dBug.zip](https://github.com/godotengine/godot/files/1747539/Camera2dBug.zip)
[Camera2dBugV2.zip](https://github.com/godotengine/godot/files/1751083/Camera2dBugV2.zip)

**Comment by eon-s** [comment]
Looks like the classic jitter issue at first glance but the effect here disappeared completely when I changed the (incorrect) use of `_process` to `_physics_process` (because it uses a `move_and_slide`).

**Comment by adeluiz** [comment]
It is not because of the note type. I've just change the note type to Sprite and using this script to move the node and the bug it is already present:
`func _process(delta):
    position  = position + Vector2(delta*elapsed,0)
    elapsed += delta*600
    pass
`

**Comment by eon-s** [comment]
Ok, then may be the camera update bug, and that was already fixed by #15060 (which may need to be cherry picked for 3.0.1, ping @akien-mga ).

Try this on your script if you want to use a workaround until the proper fix is implemented on 3.0

```gdscript
func _process(delta):
  position.x += delta*elapsed
  elapsed += delta*600
  #is like a manual camera
  var screen_size = Vector2(270, 480) #get this from viewport or settings  
  get_viewport().canvas_transform.origin = -position + screen_size / 2
```

**Comment by adeluiz** [comment]
Thanks a lot, but everything works ok if I set the smoothing speed to 0 (I just have to remember it when I create a new camera)

**Comment by adeluiz** [comment]
I've checked again, and setting the smoothing speed to 0 it works ok, so it doesn't seem related with https://github.com/godotengine/godot/pull/15060 
Looking at how it goes more off centered as the speed increased, and that doesn't happend in _physics_proccess it seems like if the update process is something as:
- the physics_proccess are updated
- the camera is update to the parent position
- the proccess are updated (and here comes the difference, and the jittering)
- everything is drawn

**Comment by KoBeWi** [comment]
I tested the reproduction project in 3.1 and seems like node gets off-centered _both_ when smoothing speed is 0 and when non-0.

**Comment by KoBeWi** [comment]
@adeluiz Can you still reproduce this in 3.2.2?

**Comment by adeluiz** [comment]
Tested on a new Windows computer with a brand new 3.2.2 stable godot and the node get off-centered with smoothing speed greater than 0 and with the workaround (speed = 0). I've even tried to put an enormous smoothing speed, but nothing changed. The camera cannot follow the sprite. Sometimes there is a jitter and the camera catches the sprite if the speed is low, but with a forced high speed on the sprite, the sprite gets off centered as it speed increase).

If I sets the sprite position on _physics_process, everything work as expected.

So if I'm not wrong there are two different options:
1. If you have to set the sprite position only on the _physics_process, please make it impossible to set on _process (throw an error or so)
2. The bug is still present since we are allowed to set the sprite position wherever we need, but camera produce jitters if you do it on _process (this is very annoying for beginners and almost made me quit from use Godot)

**Comment by KoBeWi** [comment]
> the node get off-centered with smoothing speed greater than 0 and with the workaround (speed = 0)

So that sounds like a different issue (because the original was that smoothing speed causes smoothing even when disabled).

**Comment by adeluiz** [comment]
I don't know. Maybe same bug, maybe my guess (about smoothing) was wrong. I have try something new and is funny. I have change the script on the sprite for that:
`extends Sprite


var elapsed = 800
var dir = -1
func _ready():
	# Called every time the node is added to the scene.
	# Initialization here
	pass

func _process(delta):
	position  = Vector2(dir*40  + dir * randf() * 10, 0)
	dir = -dir
	elapsed += delta*53
	pass`

That makes the sprite never on the center of the camera. But if you change the _process for _physics_process, it is always on the center

**Comment by adeluiz** [comment]
Here is the new camera jitter godot example project. 
I've added a sibling sprite, and it moves with the parent as it is expected, but the camera... the camera is never where it should be.
[Camera2dBugV3.zip](https://github.com/godotengine/godot/files/4920945/Camera2dBugV3.zip)

**Comment by eon-s** [comment]
Disabling VSync, the sprite shakes a little but stays on the center up to nearly 20k pixels/sec, then offsets a bit and start with the big jitter.

The Camera2D process mode does not change anything on the normal behavior of the OP example.

**Comment by adeluiz** [comment]
Please try the V3. It is really weird. To me it shows like everything works something like:

· all process stuff is done in a thread(atomic, protected, no render while process is done)
· render on another thread
· all physic_process stuff and camera position movement on another threat (atomic, protected)

If godot works like that, on _process there should be no movement stuff at all (only state logic or so), otherwise jitter should happens for sure, and smooth movements could be difficult to achieve. Everything that result in something that moves should be on _physic_process if I'm not wrong

**Comment by adeluiz** [comment]
I have uploaded a GIF capture of Camera2dBugV3 to see the effect of the camera being never centered on the sprite when moved with _process (remember that the camera is a sibling of the sprite, so the logic says that it has to follow the sprite on every frame, as the sibling sprite does):

![LateralScrollerV3](https://user-images.githubusercontent.com/4427432/87521014-0fb93480-c684-11ea-85e5-60f409de5601.gif)

I move the sprite to the left and right on every frame and the camera is not able to follow the sprite (the camera position seems to be updated a frame later than the sprite or the other sibling sprite). But if I change the _process for _physic_process, it is always centered (this is not shown on the gif, but you only see two sprites that doesn't seem to move on screen, always camera centered, as intended).

**Comment by KoBeWi** [comment]
Not sure what to do about this tbh. I tried the camerabugv3 and noticed that the camera has a `process_mode` property. Of all combinations, the movement isn't jittery when `process_mode` is Idle (default) and node moves in `_physics_process()`. Also calling `force_update_scroll()` on camera each frame seems to help too.

I looked into Camera2D's source and it seems to update scroll on internal process. I guess the jitter is due to desync in script process and camera process. The interesting bit is this:
https://github.com/godotengine/godot/blob/08dda79a9c0749b028743321c04bc5b1d7e94490/scene/2d/camera_2d.cpp#L225-L228
This basically means that if internal processing is disabled, the scroll updates on transform change, which means it's done just when the camera moves. And indeed, when you use `camera.set_process_internal(false)` the jitter is gone.

Either this should be documented or when smoothing is disabled we should always update scroll.

**Comment by eon-s** [comment]
Position changes in `_physics_process` is only for things that use physics bodies, it should not affect the camera behavior.

Maybe expose  `Camera.set_process_internal` on the inspector, on project settings, or bind it with the camera process mode?

**Comment by KoBeWi** [comment]
> or bind it with the camera process mode?

Looking at the code, something like `PROCESS_IMMEDIATE`, which disable all internal process, would make the most sense. The source code for smoothing and processing looks weird.

**Comment by akien-mga** [comment]
This was fixed in #46717.
